src/main/java/mesosphere/marathon/core/leadership/CandidateImpl.java

```diff
@@ -121,17 +126,38 @@ public CandidateImpl(
     public Supplier<Boolean> offerLeadership(final Leader leader)
             throws Group.JoinException, Group.WatchException, InterruptedException {
 
-        final Group.Membership membership = group.join(dataSupplier, new Command() {
-            @Override public void execute() {
-                leader.onDefeated();
-            }
-        });
+        // start a group watch, but only once in the life time of the the CandidateImpl.
+        if (this.cancelWatch == null) {
+            this.cancelWatch = this.group.watch(new Group.GroupChangeListener() {
+                @Override
+                public void onGroupChange(Iterable<String> memberIds) {
+                    Group.GroupChangeListener listener = groupChangeListener.get();
+                    if (listener != null) {
+                        listener.onGroupChange(memberIds);
+                    }
+                }
+            });
+        }
 
+        // listen for group member changes
         final AtomicBoolean elected = new AtomicBoolean(false);
         final AtomicBoolean abdicated = new AtomicBoolean(false);
-        group.watch(new Group.GroupChangeListener() {
-            @Override public void onGroupChange(Iterable<String> memberIds) {
+        final AtomicReference<Group.Membership> membershipRef = new AtomicReference<Group.Membership>(null);
+        final AtomicReference<Iterable<String>> pendingChange = new AtomicReference<Iterable<String>>(null);
+        this.groupChangeListener.set(new Group.GroupChangeListener() {
+            @Override
+            public void onGroupChange(Iterable<String> memberIds) {
                 boolean noCandidates = Iterables.isEmpty(memberIds);
+                Group.Membership membership;
+                synchronized (this) {
```

Comment:
Does this really need to be synchronized?

Quote from [ZK's doc](https://zookeeper.apache.org/doc/trunk/zookeeperProgrammers.html#Java+Binding):

> When a ZooKeeper object is created, two threads are created as well: an IO thread and an event thread. All IO happens on the IO thread (using Java NIO). All event callbacks happen on the event thread. Session maintenance such as reconnecting to ZooKeeper servers and maintaining heartbeat is done on the IO thread. Responses for synchronous methods are also processed in the IO thread. All responses to asynchronous methods and watch events are processed on the event thread. There are a few things to notice that result from this design:
>
> - All completions for asynchronous calls and watcher callbacks will be made in order, one at a time. The caller can do any processing they wish, but no other callbacks will be processed during that time.

As far as I could see, `onGroupChange` is called only when a watcher callback is triggered, but I might be missing something.